crypt.go

```diff
@@ -675,6 +675,54 @@ func setLUKS2KeyslotPreferred(devicePath string, slot int) error {
 	return nil
 }
 
+// InitializeLUKS2ContainerOptions carries options for initializing LUKS2
+// containers.
+type InitializeLUKS2ContainerOptions struct {
+	// MetadataKiBSize sets the size of the LUKS2 metadata (JSON) area,
+	// expressed in multiples of 1024 bytes. The value includes 4096 bytes
+	// for the binary metadata. According to LUKS2 specification and
+	// cryptsetup(8), only these values are valid: 16, 32, 64, 128, 256,
+	// 512, 1024, 2048 and 4096 KiB.
+	MetadataKiBSize int
+	// KeyslotsAreaSize sets the size of the LUKS2 binary keyslot area,
+	// expressed in multiples of 1024 bytes. The value must be aligned to
+	// 4096 bytes, with the maximum size of 128MB.
+	KeyslotsAreaKiBSize int
+}
+
+func validateInitializeLUKS2Options(options *InitializeLUKS2ContainerOptions) error {
+	if options == nil {
+		return nil
+	}
+
+	if options.MetadataKiBSize != 0 {
+		// metadata size is one of the allowed values (in kB)
+		allowedSizesKB := []int{16, 32, 64, 128, 256, 512, 1024, 2048, 4096}
+		found := false
+		for _, sz := range allowedSizesKB {
+			if options.MetadataKiBSize == sz {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return fmt.Errorf("cannot set metadata size to %v KiB",
+				options.MetadataKiBSize)
+		}
+	}
+	if options.KeyslotsAreaKiBSize != 0 {
+		// minimum size 4096 (4KiB), a multiple of 4096, max size 128MiB
+		sizeValid := options.KeyslotsAreaKiBSize >= 4 &&
+			options.KeyslotsAreaKiBSize <= 128*1024 &&
+			options.KeyslotsAreaKiBSize%4 == 0
+		if !sizeValid {
+			return fmt.Errorf("cannot set keyslots area size to %v KiB",
+				options.KeyslotsAreaKiBSize)
+		}
+	}
```

Comment:
No strong opinion here. The idea was to do the checks early to avoid side effects from calling cryptsetup luksFormat if any and maybe produce less obscure error messages.

Reply:
FWIW, `cryptsetup luksFormat` seems to return  a generic `Unsupported LUKS2 metadata size options.` if either of the parameters is invalid.

Reply:
then this is probably fine as is for the time being, we can change our mind later if needed.